core/src/main/scala/kafka/raft/KafkaMetadataLog.scala

```diff
@@ -29,13 +32,15 @@ import org.apache.kafka.snapshot.FileRawSnapshotReader
 import org.apache.kafka.snapshot.FileRawSnapshotWriter
 import org.apache.kafka.snapshot.RawSnapshotReader
 import org.apache.kafka.snapshot.RawSnapshotWriter
+import org.apache.kafka.snapshot.Snapshots
 
 import scala.compat.java8.OptionConverters._
 
-class KafkaMetadataLog(
+final class KafkaMetadataLog private (
```

Comment:
This PR changes the recovery logic so that the log is `truncateFullyToLatestSnapshot()` (https://github.com/apache/kafka/pull/9816/files#diff-b332f85b04775c821226b6f704e91d51f9647f29ba73dace65b99cf36f6b9ceaR298) if the `latestSnapshotId` is greater than `new OffsetAndEpoch(endOffset().offset, log.lastFetchedEpoch)`.

This can happen when a follower crashed after downloading a snapshot from the leader but before calling `truncateFullyToLatestSnapshot()`. Or https://github.com/apache/kafka/pull/9816/files#diff-b332f85b04775c821226b6f704e91d51f9647f29ba73dace65b99cf36f6b9ceaR155

I think this is correct because `KafkaRaftClient` guarantee that any offset less then high watermark has been flushed and after https://issues.apache.org/jira/browse/KAFKA-10800 any snapshot generated by the state machine will be less than or equal to the high-watermark.